### PR TITLE
Add MinimumResourceVersion option to Reflector

### PR DESCRIPTION
**Pull request by jpbetz**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Provides a way for go-client applications to provide a minimum resource version when creating reflectors to avoid https://github.com/kubernetes/kubernetes/issues/59848

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Add MinimumResourceVersion option to Reflector to allow reflectors avoid processing arbitrarily old changes when started
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [Design]: https://docs.google.com/document/d/1x3JXKwPTpum8S6bC-YXvoGS583iixuQSskEGukTAzcI/edit#
```

/sig api-machinery
/priority important-soon

### CODE REVIEW

**Comment by jpbetz** [comment]
This is ready for review.

**Comment by jpbetz** [comment]
friendly nudge

**Comment by jpbetz** [comment]
This is ready for review

**Comment by caesarxuchao** [comment]
Discussed an alternative fix with Joe.

When reflector starts, we let it issue a LIST request with rv="" and limit=0. Because rv="", it forces the apiserver to do a quorum read to get the fresh resource version, thus preventing a stale read from the watch cache. And because limit=0, it's not going to be a huge performance cost.

Then, the reflector lists with the fresh rv. This does _not_ put a heavy load on the server, because the server will wait it's watch cache to be as fresh as the rv and then serve from the cache (see [code](https://github.com/kubernetes/kubernetes/blob/14344b57e56258e87cbe80c8cd80399855eca424/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L566)).

**Comment by jpbetz** [comment]
We'll need to handle list 410 responses like described in https://github.com/kubernetes/kubernetes/pull/67998#issuecomment-538058258 for this fix too.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 1]
```diff
@@ -68,6 +68,23 @@ type Reflector struct {
 	// WatchListPageSize is the requested chunk size of initial and resync watch lists.
 	// Defaults to pager.PageSize.
 	WatchListPageSize int64
+
+	// MinimumResourceVersion sets the earliest allowed change a reflector's initial list and watch may start from.
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: I'd suggest avoiding language that implies all values must be numbers. Something like `OldestAllowedResourceVersion` might be better (though I don't love the name).

**Comment by caesarxuchao** [comment]
Does https://github.com/kubernetes/kubernetes/pull/82862#issuecomment-538134538 make sense?

**Comment by jpbetz** [comment]
> Does #82862 (comment) make sense?

@caesarxuchao I realized we can simplify further. Since reflectors use pagination which already skips the watch cache for list requests (https://github.com/kubernetes/kubernetes/blob/25a728ae5eb0ae067d21679fad915c0555242470/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L591), we can simply have the initial list request of a reflector use RV="" to start the reflector at the latest RV.

That is simple enough that I can do it in #83520. I'll close this one. Thanks for the help!

**Comment by liggitt** [comment]
> Since reflectors use pagination

reflectors do *not* use pagination today. see https://groups.google.com/forum/#!msg/kubernetes-sig-api-machinery/SfCIk5MccOc/_bKbrCsoAwAJ

**Comment by liggitt** [comment]
hrmm... am I mixing streams? https://github.com/kubernetes/kubernetes/blob/release-1.16/staging/src/k8s.io/client-go/tools/cache/reflector.go#L179-L189 shows reflectors implementing paging logic, but I can't find any non-test instances of callers setting WatchListPageSize to trigger paging

**Comment by jpbetz** [comment]
The default `WatchListPageSize` should be the default page size of `500`.

refs: https://github.com/kubernetes/kubernetes/blob/release-1.16/staging/src/k8s.io/client-go/tools/cache/reflector.go#L184-L186, https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/pager/pager.go#L77-L79

**Pull request closed without merging**